File: oxidize-golang/internal/cli/gpucluster.go
package cli

import (
	"flag"
	"fmt"
	"io"

	"github.com/Zapdev-labs/oxidize/golang/core/gpucluster"
)

func gpuClusterCommand(args []string, stdout, stderr io.Writer) error {
	sub := "help"
	if len(args) > 0 {
		sub = args[0]
	}
	switch sub {
	case "profiles":
		for _, p := range gpucluster.AllProfiles() {
			_, _ = fmt.Fprintf(stdout,
				"%-14s product=%-26s arch=%-9s mem=%dMiB tdp=%dW nvlink=%v mig=%v timeslice=%d net=%s\n",
				p.Family.Slug(), p.Product, p.Generation, p.MemoryMiB, p.TDPWatts,
				p.NVLink, p.MIGCapable, p.TimeSliceReplicas, p.NetworkClass)
		}
		return nil
	case "detect":
		gpus := gpucluster.DetectGPUs()
		if len(gpus) == 0 {
			_, _ = fmt.Fprintln(stdout, "no NVIDIA GPUs detected (nvidia-smi unavailable or no devices)")
			return nil
		}
		for _, g := range gpus {
			fam := "unknown"
			if g.FamilyKnown {
				fam = g.Family.Slug()
			}
			_, _ = fmt.Fprintf(stdout, "GPU %d: %s (%dMiB) family=%s mig=%v\n",
				g.Index, g.Name, g.MemoryTotalMiB, fam, g.MIGEnabled)
		}
		_, _ = fmt.Fprintln(stdout, "--- summary ---")
		for _, c := range gpucluster.Summarize(gpus) {
			_, _ = fmt.Fprintf(stdout, "%s: %d\n", c.Family.Slug(), c.Count)
		}
		return nil
	case "generate":
		return gpuClusterGenerate(args[1:], stdout, stderr)
	default:
		_, _ = fmt.Fprintln(stdout,
			"usage: oxidize gpu-cluster <generate|detect|profiles>\n\n"+
				"generate [--family b200|a100|rtx-pro-6000] [--nodes N] [--gpus-per-node N]\n"+
				"detect   probe local NVIDIA GPUs via nvidia-smi\n"+
				"profiles list known GPU tier profiles")
		return nil
	}
}

func gpuClusterGenerate(args []string, stdout, stderr io.Writer) error {
	fs := flag.NewFlagSet("gpu-cluster generate", flag.ContinueOnError)
	fs.SetOutput(io.Discard)
	familyStr := fs.String("family", "", "b200|a100|rtx-pro-6000 (default: full three-tier cluster)")
	nodes := fs.Uint("nodes", 0, "node count")
	gpusPerNode := fs.Uint("gpus-per-node", 0, "physical GPUs per node")
	if err := fs.Parse(args); err != nil {
		return err
	}

	var specs []gpucluster.NodePoolSpec
	if *familyStr != "" {
		fam, ok := gpucluster.FamilyFromSlug(*familyStr)
		if !ok {
			_, _ = fmt.Fprintln(stderr, "error: --family expects b200|a100|rtx-pro-6000")
			return fmt.Errorf("invalid family %q", *familyStr)
		}
		count := uint32(*nodes)
		if count == 0 {
			count = defaultNodeCount(fam)
		}
		gpn := uint32(*gpusPerNode)
		if gpn == 0 {
			gpn = defaultGPUsPerNode(fam)
		}
		specs = []gpucluster.NodePoolSpec{{Family: fam, NodeCount: count, GPUPerNode: gpn}}
	} else {
		specs = []gpucluster.NodePoolSpec{
			{Family: gpucluster.B200, NodeCount: 8, GPUPerNode: 8},
			{Family: gpucluster.A100, NodeCount: 16, GPUPerNode: 8},
			{Family: gpucluster.RTXPro6000, NodeCount: 4, GPUPerNode: 2},
		}
	}

	families := make([]gpucluster.Family, 0, len(specs))
	for _, s := range specs {
		families = append(families, s.Family)
	}

	_, _ = fmt.Fprint(stdout, gpucluster.NodePoolsYAML(specs))
	_, _ = fmt.Fprintln(stdout, "---")
	_, _ = fmt.Fprint(stdout, gpucluster.DevicePluginConfigYAML(families))
	for _, f := range families {
		if mig, ok := gpucluster.MIGConfigYAML(f); ok {
			_, _ = fmt.Fprintln(stdout, "---")
			_, _ = fmt.Fprint(stdout, mig)
		}
	}
	_, _ = fmt.Fprintln(stdout, "---")
	_, _ = fmt.Fprint(stdout, gpucluster.PrometheusRulesYAML())
	for _, f := range families {
		_, _ = fmt.Fprintln(stdout, "---")
		_, _ = fmt.Fprint(stdout, gpucluster.HelmValuesYAML(f))
	}
	return nil
}

func defaultNodeCount(f gpucluster.Family) uint32 {
	switch f {
	case gpucluster.B200:
		return 8
	case gpucluster.A100:
		return 16
	case gpucluster.RTXPro6000:
		return 4
	default:
		return 1
	}
}

func defaultGPUsPerNode(f gpucluster.Family) uint32 {
	switch f {
	case gpucluster.RTXPro6000:
		return 2
	default:
		return 8
	}
}

File: oxidize-cli/src/main/gpu_cluster.rs
/// Handle the `oxidize gpu-cluster <subcommand>` family.
///
/// Subcommands:
///   generate [--family b200|h100|a100|rtx-pro-6000] [--nodes N] [--gpus-per-node N]
///            Emit the Kubernetes/Helm manifests from the GPU cluster spec.
///   detect   Probe the local node for NVIDIA GPUs via nvidia-smi.
///   profiles List the known GPU tier profiles.
pub(super) fn run_gpu_cluster(args: &[String]) -> i32 {
    use oxidize_core::gpu_cluster as gc;

    let sub = args.first().map(String::as_str).unwrap_or("help");
    match sub {
        "profiles" => {
            for p in gc::all_profiles() {
                println!(
                    "{:<14} product={:<26} arch={:<9} mem={}MiB tdp={}W nvlink={} mig={} timeslice={} net={}",
                    p.family.slug(),
                    p.product,
                    p.generation,
                    p.memory_mib,
                    p.tdp_watts,
                    p.nvlink,
                    p.mig_capable,
                    p.time_slice_replicas,
                    p.network_class,
                );
            }
            0
        }
        "detect" => {
            let gpus = gc::detect_gpus();
            if gpus.is_empty() {
                println!("no NVIDIA GPUs detected (nvidia-smi unavailable or no devices)");
                return 0;
            }
            for g in &gpus {
                let fam = g.family.map(|f| f.slug()).unwrap_or("unknown");
                println!(
                    "GPU {}: {} ({}MiB) family={} mig={}",
                    g.index, g.name, g.memory_total_mib, fam, g.mig_enabled
                );
            }
            println!("--- summary ---");
            for (fam, n) in gc::summarize(&gpus) {
                println!("{}: {}", fam.slug(), n);
            }
            0
        }
        "generate" => {
            let mut family: Option<gc::GpuFamily> = None;
            let mut nodes: u32 = 0;
            let mut gpus_per_node: u32 = 0;
            let mut i = 1;
            while i < args.len() {
                match args[i].as_str() {
                    "--family" => {
                        i += 1;
                        match args.get(i).and_then(|v| gc::GpuFamily::from_slug(v)) {
                            Some(f) => family = Some(f),
                            None => {
                                eprintln!("error: --family expects b200|h100|a100|rtx-pro-6000");
                                return 2;
                            }
                        }
                    }
                    "--nodes" => {
                        i += 1;
                        match args.get(i) {
                            Some(v) => match v.parse() {
                                Ok(n) => nodes = n,
                                Err(_) => {
                                    eprintln!(
                                        "error: --nodes expects a positive integer, got '{v}'"
                                    );
                                    return 2;
                                }
                            },
                            None => {
                                eprintln!("error: --nodes requires a value");
                                return 2;
                            }
                        }
                    }
                    "--gpus-per-node" => {
                        i += 1;
                        match args.get(i) {
                            Some(v) => match v.parse() {
                                Ok(n) => gpus_per_node = n,
                                Err(_) => {
                                    eprintln!(
                                        "error: --gpus-per-node expects a positive integer, got '{v}'"
                                    );
                                    return 2;
                                }
                            },
                            None => {
                                eprintln!("error: --gpus-per-node requires a value");
                                return 2;
                            }
                        }
                    }
                    other => {
                        eprintln!("error: unknown flag {other}");
                        return 2;
                    }
                }
                i += 1;
            }

            // Default to the full three-tier cluster from the spec when no
            // single family is selected.
            let specs = match family {
                Some(f) => {
                    let count = if nodes > 0 {
                        nodes
                    } else {
                        default_node_count(f)
                    };
                    let gpn = if gpus_per_node > 0 {
                        gpus_per_node
                    } else {
                        default_gpus_per_node(f)
                    };
                    vec![gc::NodePoolSpec::new(f, count, gpn)]
                }
                None => vec![
                    gc::NodePoolSpec::new(gc::GpuFamily::B200, 8, 8),
                    gc::NodePoolSpec::new(gc::GpuFamily::H100, 4, 8),
                    gc::NodePoolSpec::new(gc::GpuFamily::A100, 16, 8),
                    gc::NodePoolSpec::new(gc::GpuFamily::RtxPro6000, 4, 2),
                ],
            };
            let families: Vec<gc::GpuFamily> = specs.iter().map(|s| s.family).collect();

            print!("{}", gc::node_pools_yaml(&specs));
            println!("---");
            print!("{}", gc::device_plugin_config_yaml(&families));
            for f in &families {
                if let Some(mig) = gc::mig_config_yaml(*f) {
                    println!("---");
                    print!("{mig}");
                }
            }
            println!("---");
            print!("{}", gc::prometheus_rules_yaml());
            for f in &families {
                println!("---");
                print!("{}", gc::helm_values_yaml(*f));
            }
            0
        }
        _ => {
            eprintln!(
                "usage: oxidize gpu-cluster <generate|detect|profiles>\n\
                 \n\
                 generate [--family b200|h100|a100|rtx-pro-6000] [--nodes N] [--gpus-per-node N]\n\
                 detect   probe local NVIDIA GPUs via nvidia-smi\n\
                 profiles list known GPU tier profiles"
            );
            1
        }
    }
}

pub(super) fn default_node_count(f: oxidize_core::gpu_cluster::GpuFamily) -> u32 {
    use oxidize_core::gpu_cluster::GpuFamily::*;
    match f {
        B200 => 8,
        H100 => 4,
        A100 => 16,
        RtxPro6000 => 4,
    }
}

pub(super) fn default_gpus_per_node(f: oxidize_core::gpu_cluster::GpuFamily) -> u32 {
    use oxidize_core::gpu_cluster::GpuFamily::*;
    match f {
        B200 | H100 | A100 => 8,
        RtxPro6000 => 2,
    }
}
